### DISCUSSION TITLE
Some variables are empty in project context

### DISCUSSION THREAD

**Discussion by AKhrulev** [wixtoolset/issues#7462] [Questions]
I use post-build event in this way:

```
<Project Sdk="WixToolset.Sdk/4.0.0">
  <PropertyGroup>
    <Configurations>Release Pro;Release Std</Configurations>
    <Platforms>x64</Platforms>
    <OutputName>MyProjectName</OutputName>
    <PostBuildEvent>xcopy "$(ProjectDir)$(OutDir)$(TargetFileName)" "$(ProjectDir)binToSign\$(Platform)\$(Configuration)\" /Y</PostBuildEvent>
    <RunPostBuildEvent>OnOutputUpdated</RunPostBuildEvent>
  </PropertyGroup>
...
```
But the result in Output is:

`error MSB3073: The command "xcopy "" "binToSign\x64\Release Pro\" /Y" exited with code 4.`

We see that `$(Platform)` and `$(Configuration)` are valid, but variables `$(ProjectDir)`, `$(OutDir)`, `$(TargetFileName)` are empty. How to have correct values of these empty variables in project context?

WiX Toolset Visual Studio Extension version 1.0.0.22
wix version: 4.0.0+8c757c0f

**Comment by BMurri** [comment]
This has nothing to do with WiX per se. It is a classic `MSBuild` "issue" that will never be "fixed". The issue is that root-level `PropertyGroup` (and `ItemGroup`) elements in MSBuild files are fully evauated at the time they are parsed, and the properties you are depending on haven't been defined above their use (or aren't defined at the root-level of any msbuild file or argument). You will find that an equivalent property in a C# project will fail in the same way.

The classic solution is to move your definition to a target that runs before the target that consumes that property (or item, as the case may be).

**Comment by robmen** [comment]
`PostBuildEvent` is an anachronism in Sdk-style projects. You can do it, but it requires using explicit Sdk imports. HeatWave will automatically convert v3 .wixprojs using `PostBuildEvent` to use explicit Sdk imports, but if you manually add `PostBuildEvent` in a v4 .wixproj, you'll need to do the manual Sdk import yourself.

This topic has been discussed on this Discussion forum in the past for more information.
